src/compiler/parser.rs

```diff
@@ -0,0 +1,306 @@
+
+        self.set_file_name(&normalize_path(_file_name));
+
+        self.set_parse_diagnostics(vec![]);
+
+        self.scanner.set_text(Some(_source_text), None, None);
+    }
+
+    fn parse_source_file_worker(&mut self) -> SourceFile {
+        self.next_token();
+
+        let statements =
+            self.parse_list(ParsingContext::SourceElements, ParserType::parse_statement);
+        Debug_.assert(matches!(self.token(), SyntaxKind::EndOfFileToken));
+
+        let source_file = self.create_source_file(self.file_name(), statements);
+
+        source_file
+    }
+
+    fn create_source_file<TNodes: Into<NodeArrayOrVec>>(
+        &self,
+        _file_name: &str,
+        statements: TNodes,
+    ) -> SourceFile {
+        self.factory.create_source_file(self, statements)
+    }
+
+    fn parse_error_at_current_token(&mut self, message: &DiagnosticMessage) {
+        self.parse_error_at(
+            self.scanner.get_token_pos(),
+            self.scanner.get_text_pos(),
+            message,
+        );
+    }
+
+    fn parse_error_at_position(
+        &mut self,
+        start: usize,
+        length: usize,
+        message: &DiagnosticMessage,
+    ) {
+        let last_error = last_or_undefined(self.parse_diagnostics());
+        if match last_error {
+            None => true,
+            Some(last_error) => last_error.start != start,
+        } {
+            let file_name = self.file_name().to_string();
+            self.parse_diagnostics().push(create_detached_diagnostic(
+                &file_name, start, length, message,
+            ));
+        }
+
+        self.parse_error_before_next_finished_node = true;
+    }
+
+    fn parse_error_at(&mut self, start: usize, end: usize, message: &DiagnosticMessage) {
+        self.parse_error_at_position(start, end - start, message);
+    }
+
+    fn token(&self) -> SyntaxKind {
+        self.current_token()
+    }
+
+    fn next_token_without_check(&mut self) -> SyntaxKind {
+        let current_token = self.scanner.scan();
+        self.set_current_token(current_token);
+        self.current_token()
+    }
+
+    fn next_token(&mut self) -> SyntaxKind {
+        self.next_token_without_check()
+    }
+
+    fn is_identifier(&self) -> bool {
+        false
+    }
+
+    fn parse_expected(&mut self, kind: SyntaxKind, should_advance: Option<bool>) -> bool {
+        let should_advance = should_advance.unwrap_or(true);
+        if is_same_variant(&self.token(), &kind) {
+            if should_advance {
+                self.next_token();
+            }
+            return true;
+        }
+
+        self.parse_error_at_current_token(&Diagnostics::_0_expected);
+        false
+    }
+
+    fn create_node_array(&self, elements: Vec<Node>) -> NodeArray {
+        self.factory.create_node_array(elements)
+    }
+
+    fn finish_node<TParsedNode: NodeInterface>(&mut self, node: TParsedNode) -> TParsedNode {
```

Comment:
So `Node` is now a concrete enum type, but sort of per above comment re shared "base node" fields, there are places where we want to treat whichever concrete node subtype "as a node" (eg to manipulate its shared "base node" fields). So I exposed `NodeInterface` to represent that (so eg `EmptyStatement` implements `NodeInterface`), which would consist of getters + setters for those base node fields